### PR TITLE
[no squash] Implement tool use sound

### LINKED ISSUE #11586: Add support for a sound to be played on tool use
## Problem

The only way to easily play a sound on tool use is to use the on_use function, the problem with this is that it overrides the basic uses of the tool and only does what you define inside the function.

## Solutions

I would suggest either another item definition function for playing a sound on_use as not to override the basic actions, 
Or to add an event under the [sound](https://pastebin.com/gUY1fLDV) function such as "hits = \<SimpleSoundSpec\>"

## Example

An example of the current on_use function can be seen here, this works for playing a sound when the lmb is clicked but again, it overrides the basic uses of the tool such as attacking, digging, or picking up items.

```lua
minetest.register_tool("example:tool", {
     description =  "tool",
     inventory_image = "tool",
     wield_image = "tool",
     stack_max = 1,
     tool_capabilities = {
        full_punch_interval = 1.0,
        max_drop_level = 0,
        groupcaps = {
            -- For example:
            choppy = {times = {[1] = 2.50, [2] = 1.40, [3] = 1.00},
                     uses = 128, maxlevel = 2},
        },
        damage_groups = {fleshy=5},
        punch_attack_uses = 256,
    },
   
    on_use = function(itemstack, user, pointed_thing)
      minetest.sound_play(
        {name = "default_place_node"},
        {
          object = user,
          gain = 1.0,  -- default
          max_hear_distance = 32,  -- default, uses an euclidean metric
          loop = false,
        },
        true
      )
      return
  })
```

### PR DESCRIPTION
**Pull request by sfan5**
fixes #11586

~~<b>Open question</b>:  When digging using a tool with a sound, should the sound play for each swing or on the first one?
I initially thought only the first, but each swing is easier to get right and probably makes more sense anyway.~~

## To do

This PR is Ready for Review.

## How to test

devtest `soundstuff:punch_use` and `soundstuff:punch_use_air`
try to punch:
* a node (digging)
* an entity
* nothing

### CODE REVIEW

File: src/client/game.cpp [hunk 1]
```diff
@@ -3698,38 +3699,31 @@ void Game::handleDigging(const PointedThing &pointed, const v3s16 &nodepos,
...
+				player, nodepos, n, features);
 		}
 
 
 		// Send event to trigger sound
-		client->getEventManager()->put(new NodeDugEvent(nodepos, wasnode));
+		client->getEventManager()->put(new NodeDugEvent(nodepos, n));
 	}
 
 	if (runData.dig_time_complete < 100000.0) {
```
Comment (sfan5) [inline src/client/game.cpp:3729]: In the end I didn't need to touch this function but I cleaned it up anyway.

**Comment by Wuzzy2** [comment]
Interesting. I wonder if you could implement hand swooshes with this as well. And that the "hand swooshes" ONLY apply to the hand and nothing else


This could be used for sword wooshes. I'm not sure if I like the fact that the same sound is used for all targets (entity, nothing, node). This can be alleviated by adding more sound variants (`*.1.ogg`, `*.2.ogg`, etc.) but I'm not sure if you can make it sound good. A quick test with some dummy woosh sounds and I notice the sound gets repetitive *fast*. But maybe this can be solved by simply using high-quality sounds, but I'm not sure about that.

Anyway, I'm just thinking out aloud. But I think it's also important to think about the gameplay implications and not just purely look at this from a technical perspective.

Maybe it makes more sense to separate this into a generic sound for any target, and a sound specifically for punching nothing. The latter still gives games the chance to customize sounds within the entity/node on_punch handler (or whatever) can do custom sounds. This is still ugly as those things now need to know the item that hits them, but at least it works. The former solution clearly leaves no such flexibility, as there is no way you could do clean per-target sword hit sounds when you have `use` specified. The `use` sound will alway interfere but if you remove the `use` sound, you no longer have an "air woosh".
Therefore, I think it might be a good idea to have:
1) sound for punching nothing
2) sound for punching something (anything that is not nothing)

That way, the game designer can decide what to do. If the game designer wants to have the sound played for all punches (your current implementation), they just have to set no. 1 and 2 to equal sounds.

Sadly, the sound interval / the interval in which the item "swings" is hardcoded. But this is a different feature, ignore it. At least the sound is in sync with the animation, which is good.

I dislike the `use` identifier, this is a weasel word and Minetest has been abusing the word "use" in the API for various different things in a confusing way already, let's not make it worse. Maybe say `punch_use` to make it clear it is *that* kind of use. Or even `punch_use_item` because in theory you could also argue a `punch_use` might apply to a node.

Please set `uses=0` in the DevTest tool definition, so the tool doesn't break.

> Open question: When digging using a tool with a sound, should the sound play for each swing or on the first one?

Maybe just impement both? I think both use cases sound useful. Could be split into `sound_on_first_swing` and `sound_on_subsequent_swings`. Again, the game designer can set both equal if they want constant swinging sounds.

Final question: Does this sound still play when the tool has an `on_use` handler?

**Comment by sfan5** [comment]
> I wonder if you could implement hand swooshes with this as well. And that the "hand swooshes" ONLY apply to the hand and nothing else

Sure, the hand is just another item.

> Sadly, the sound interval / the interval in which the item "swings" is hardcoded.

Isn't that equal to the `full_punch_interval`?

> Maybe it makes more sense to separate this into a generic sound for any target, and a sound specifically for punching nothing.

That's a good idea, will do.

> Final question: Does this sound still play when the tool has an on_use handler?

Yes.

**Comment by Desour** [comment]
> > Sadly, the sound interval / the interval in which the item "swings" is hardcoded.
> 
> Isn't that equal to the `full_punch_interval`?

`full_punch_interval` is only used when punching objects, not when digging nodes or air.

**Comment by sfan5** [comment]
So, does anyone think this is useful?

**Comment by Zughy** [comment]
@sfan5 I think it is, it adds immersion. I agree with @Wuzzy2 's idea of adding multiple optional sounds to the same tool (maybe in a future PR, it's not that necessary as for now), in order to randomise them every time the tool is used (like in the sword example).  
To provide another usecase, it'd work on a gun. However, if the sound is always reproduced, it can only work on infinite ammo guns that never need to reload. How about not reproducing the sound if `on_use` explicitly returns `false` or if the interval between an use and another wasn't reached? Just spitballing

> When digging using a tool with a sound, should the sound play for each swing or on the first one?

The former makes more sense to me

**Comment by Wuzzy2** [comment]
> in order to randomise them every time the tool is used (like in the sword example).

Isn't this already automatically working with the `sound.1.ogg`, `sound.2.ogg`, `sound.3.ogg`, etc. naming scheme that is used for all sounds already?

**Review by Df458 — COMMENTED** [review]
Half-approval. I've taken an initial look at this and it looks good, but due to some health issues I won't be able to test further for the time being.

File: src/script/common/c_content.cpp [hunk 2]
```diff
@@ -112,6 +112,19 @@ void read_item_definition(lua_State* L, int index,
 	}
 	lua_pop(L, 1);
 
+	// No, this is not a mistake. Item sounds are in "sound", node sounds in "sounds".
```
[the remark this answers is not in the document]
Comment (sfan5) [reply src/script/common/c_content.cpp:115]: I'm planning a later PR to add lots of stuff to the breakages file, will keep in mind.

File: src/itemdef.cpp [hunk 3]
```diff
@@ -214,12 +221,15 @@ void ItemDefinition::deSerialize(std::istream &is, u16 protocol_version)
...
-	// If you add anything here, insert it primarily inside the try-catch
+	// If you add anything here, insert it inside the try-catch
 	// block to not need to increase the version.
 	try {
 		short_description = deSerializeString16(is);
 
 		place_param2 = readU8(is); // 0 if missing
+
+		sound_use.deSerialize(is, CONTENTFEATURES_VERSION);
+		sound_use_air.deSerialize(is, CONTENTFEATURES_VERSION);
```
Comment (SmallJoker) [inline src/itemdef.cpp:232]:
```suggestion
		sound_use.deSerialize(is, protocol_version);
		sound_use_air.deSerialize(is, protocol_version);
```
Rebase issue caused by versioning changes in the meantime (although it does functionally not matter as of now).

Comment (sfan5) [reply src/itemdef.cpp:232]: hmm I'm pretty sure I fixed this conflict

File: src/client/game.cpp [hunk 4]
```diff
@@ -249,23 +249,23 @@ class NodeDugEvent: public MtEvent
...
+
 public:
 	bool makes_footstep_sound;
 	float m_player_step_timer;
 	float m_player_jump_timer;
 
 	SimpleSoundSpec m_player_step_sound;
 	SimpleSoundSpec m_player_leftpunch_sound;
+	// Second sound made on left punch, currently used for item 'use' sound
+	SimpleSoundSpec m_player_leftpunch_sound2;
```
Comment (rubenwardy) [inline src/client/game.cpp:268]: perhaps `m_player_leftpunch_sound_wield` and `m_player_leftpunch_sound_player` or something, to disambiguate them

**Review by rubenwardy — APPROVED** [review]

**Comment by rubenwardy** [comment]
tested

**Pull request merged**